### PR TITLE
Implementing vscale_range with msve-vector-bit into Flang

### PR DESCRIPTION
**Pull request by abrahamtovarmob**
depends on https://github.com/flang-compiler/flang/pull/1173
depends on https://github.com/flang-compiler/flang/pull/1186

Implement msve-vector-bits in classic Flang. Implementing vscale_range with the functionallity to read from msve-vector-bits to be able to manipulate the range of the scale. for vscale_range(X/128, X/128)

Important to add the previous commits since the base of this work need those changes too.

### CODE REVIEW

File: test/llvm_ir_correct/attr.f90 [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+      program tz
+       integer :: i
+       integer ::acc(100)
+       do i=1,100
+            acc(i) = 5
+       end do
+      print *, acc(100)
+      end program
+! ATTRS1: attributes{{.*}}"target-features"="+neon"
+! ATTRS2: attributes{{.*}}"target-features"="+neon,+sve"
```
Comment (pawosm-arm) [inline test/llvm_ir_correct/attr.f90]: Newline missing after the last line.

**Review by pawosm-arm — CHANGES_REQUESTED** [review]

File: tools/flang2/flang2exe/cgmain.cpp [hunk 2]
```diff
@@ -13567,6 +13568,13 @@ print_function_signature(int func_sptr, const char *fn_name, LL_ABI_Info *abi,
     print_token(" \"target-features\"=\"");
     print_token(flg.target_features);
     print_token("\"");
+    int ret = extract_arch_feat(flg.target_features, mod);
+    if(ret) {
```
Comment (pawosm-arm) [inline tools/flang2/flang2exe/cgmain.cpp:13572]: `if (ret) {` it should be

File: tools/flang2/flang2exe/cgmain.cpp [hunk 3]
```diff
@@ -13567,6 +13568,13 @@ print_function_signature(int func_sptr, const char *fn_name, LL_ABI_Info *abi,
...
+      ret = is_vscale_feat(mod);
+      if(ret) {
```
Comment (pawosm-arm) [inline tools/flang2/flang2exe/cgmain.cpp:13574]: same as above

File: test/llvm_ir_correct/vscale-mbits.f90 [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      end program
+
+! ATTRS1: attributes #{{[0-9]+}}
+! ATTRS1-DAG:"target-features"="+neon,+sve"
+! ATTRS1-DAG:vscale_range(1,1)
+! ATTRS2: attributes #{{[0-9]+}}
+! ATTRS2-DAG:"target-features"="+neon,+sve"
+! ATTRS2-DAG:vscale_range(2,2)
+! ATTRS3: attributes #{{[0-9]+}}
+! ATTRS3-DAG:"target-features"="+neon,+sve2"
```
Comment (pawosm-arm) [inline test/llvm_ir_correct/vscale-mbits.f90:25]: `-march=armv8-a+sve2` implies "+neon,+sve2,+sve", this check will not pass.

The same goes to lines below.

File: tools/flang2/flang2exe/cgmain.cpp [hunk 5]
```diff
@@ -13598,6 +13606,78 @@ INLINE void static add_property_struct(char *func_name, int nreductions,
...
+      //flg.min = argument1;
+      //flg.max = argument2;
+      break;
+    }
+  }
+
+  return result;
+}
+
+bool
```
Comment (pawosm-arm) [inline tools/flang2/flang2exe/cgmain.cpp:13655]: There's something wrong with this function, it spilled mod equals to `+sve2,+sve` which resulted in `is_vscale_feat()` couldn't recognize `+sve2` in simple `strcmp()`.

**Review by pawosm-arm — CHANGES_REQUESTED** [review]
It won't work.

File: tools/flang2/flang2exe/cgmain.cpp [hunk 6]
```diff
@@ -13598,6 +13606,78 @@ INLINE void static add_property_struct(char *func_name, int nreductions,
...
+  if(i==len) {
+    result = false;
+    return result;
+  }
+
+  //here is the feature that will be extracted
+  for(j=0; i<len; ++j, ++i) {
+      str1[j] = arch_feat[i];
+  }
+  mod = (char*) calloc(j+1,sizeof(char));
```
Comment (pawosm-arm) [inline tools/flang2/flang2exe/cgmain.cpp:13684]: who frees that?!

File: tools/flang2/flang2exe/cgmain.cpp [hunk 7]
```diff
@@ -13598,6 +13606,78 @@ INLINE void static add_property_struct(char *func_name, int nreductions,
...
+  print_token(" \"vscale-range\"=\"");
+  print_token(token);
+  print_token("\"");
+}
+
+bool
+is_vscale_feat(char *mod)
+{
+  bool result = false;
+  int len = sizeof(*CPU_VSCALE)-1;
```
Comment (pawosm-arm) [inline tools/flang2/flang2exe/cgmain.cpp:13623]: that's not how sizeof should be used! `CPU_VSCALE` is an array of pointers, so a single element of it (`*CPU_VSCALE`) is a pointer, on 64-bit machine it is 8 bytes in size, so sizeof returns always 8 here. You were lucky that `CPU_VSCALE` contains 7 elements and that `8 - 1` is also 7.

File: tools/flang2/flang2exe/cgmain.cpp [hunk 8]
```diff
@@ -13598,6 +13606,78 @@ INLINE void static add_property_struct(char *func_name, int nreductions,
...
+{
+  char token[60];
+  sprintf(token, "%d,%d", flg.min, flg.max);
+  print_token(" \"vscale-range\"=\"");
+  print_token(token);
+  print_token("\"");
+}
+
+bool
+is_vscale_feat(char *mod)
```
Comment (pawosm-arm) [inline tools/flang2/flang2exe/cgmain.cpp:13620]: should take `const char *mod` instead. The `mod` variable is not modified by this function.

**Comment by pawosm-arm** [comment]
I need to make a final blow to this PR: flang2 should remain architecture independent, hence vscale_range computation depending on particular VLA implementations should not be computed here. It should be computed in LLVM's frontent driver and passed to flang2 as yet another command line option. Emission of vscale_range should therefore depend on presence of that option.

**Review by jdomke — COMMENTED** [review]
Sorry to jump in, but I tested this PR and found multiple issues in tools/flang2/flang2exe/cgmain.cpp:

- str1[30] in extract_arch_feat is too small for a64fx which yields '+v8.2a,+fp-armv8,+neon,+crc,+crypto,+fullfp16,+ras,+lse,+rdm,+sve,+sha2,+aes' in arch_feat string
- extract_arch_feat extracts everything after the 1st comma
- is_vscale_feat assumes single feature token, but gets multiple comma-separed tokens, hence strcmp fails to find anything unless you are lucky and +sve is in 2nd position of original string
- provided test cases are insufficiant since they all yield a similar '+neon,+sve..' arch_feat string and hence don't find any of the errors which I ran into with realistic march/mcpu flags

**Comment by pawosm-arm** [comment]
A decision was made that this feature needs to be provided by a new PR introducing different implementation shifting more of the logic to the frontend driver.

**Pull request closed without merging**